File: problems/arrays_and_hashing/cpp/283_move_zeroes.cpp
#include <iostream>
#include <vector>
using namespace std;

// Time Complexity: O(n) - single pass through the array
// Space Complexity: O(1) - only uses two pointers, no extra space
class Solution {
public:
  void moveZeroes(vector<int> &nums) {
    int left = 0; // pointer to place next non-zero element

    for (int right = 0; right < nums.size(); ++right) {
      if (nums[right] != 0) {
        swap(nums[left++], nums[right]);
      }
    }
  }
};

File: problems/arrays_and_hashing/rust/283_move_zeroes.rs
/ Time Complexity: O(n) - single pass through the array
// Space Complexity: O(1) - only uses two pointers, no extra space

impl Solution {
    pub fn move_zeroes(nums: &mut Vec<i32>) {
        let mut left = 0;
        
        for right in 0..nums.len() {
            if nums[right] != 0 {
                nums.swap(left, right);
                left += 1;
            }
        }
    }
}